File: two_values_solution.py
"""Exact solver for the variant with only values 1 and 2.

Input:  n
Output: maximum possible sum and one optimal n by n arrangement

A 2-cell must have a four-directionally adjacent 1-cell.  The frontier DP
minimizes the number of 1-cells; the maximum sum is then 2*n*n minus that
minimum. Parent transitions are retained to reconstruct one optimum.
"""

from __future__ import annotations

import sys


# Frontier state for one processed cell:
#   0: the cell is 1
#   1: the cell is 2 and is already adjacent to a 1
#   2: the cell is 2 and still needs a 1 on its right or below


def solve(n: int) -> tuple[int, list[str]]:
    if n <= 0:
        raise ValueError("n must be positive")

    power3 = [1] * (n + 1)
    for i in range(1, n + 1):
        power3[i] = 3 * power3[i - 1]

    # State 1 is neutral for the virtual cells above the first row: it does
    # not require domination and does not falsely dominate the first row.
    initial_state = sum(power3[column] for column in range(n))
    dp: dict[int, int] = {initial_state: 0}
    parents: list[dict[int, tuple[int, int]]] = []

    for _row in range(n):
        for column in range(n):
            next_dp: dict[int, int] = {}
            step_parents: dict[int, tuple[int, int]] = {}
            current_power = power3[column]
            left_power = 0 if column == 0 else power3[column - 1]

            for state, old_ones in dp.items():
                above = (state // current_power) % 3
                left = 1 if column == 0 else (state // left_power) % 3

                # Put 1 in the current cell.  It satisfies an undominated
                # cell above and changes an undominated left cell to state 1.
                new_state = state - above * current_power
                if column > 0 and left == 2:
                    new_state -= left_power

                new_cost = old_ones + 1
                previous = next_dp.get(new_state)
                if previous is None or new_cost < previous:
                    next_dp[new_state] = new_cost
                    step_parents[new_state] = (state, 1)

                # Put 2 in the current cell.  This is impossible if the cell
                # above still needs a 1, since all its neighbors are decided.
                if above != 2:
                    dominated = above == 0 or (column > 0 and left == 0)
                    new_digit = 1 if dominated else 2
                    new_state = state + (new_digit - above) * current_power

                    previous = next_dp.get(new_state)
                    if previous is None or old_ones < previous:
                        next_dp[new_state] = old_ones
                        step_parents[new_state] = (state, 2)

            dp = next_dp
            parents.append(step_parents)

    best = n * n + 1
    best_state = -1
    for state, number_of_ones in dp.items():
        encoded = state
        valid = True
        for _column in range(n):
            if encoded % 3 == 2:
                valid = False
                break
            encoded //= 3
        if valid:
            if number_of_ones < best:
                best = number_of_ones
                best_state = state

    # Follow the chosen transitions backward to recover one optimal grid.
    grid = [["?"] * n for _ in range(n)]
    state = best_state
    for step in range(n * n - 1, -1, -1):
        previous_state, value = parents[step][state]
        row, column = divmod(step, n)
        grid[row][column] = str(value)
        state = previous_state

    maximum = 2 * n * n - best
    return maximum, ["".join(row) for row in grid]


def minimum_number_of_ones(n: int) -> int:
    maximum, _grid = solve(n)
    return 2 * n * n - maximum



def maximum_sum(n: int) -> int:
    maximum, _grid = solve(n)
    return maximum


def main() -> None:
    data = sys.stdin.buffer.read().split()
    if len(data) != 1:
        raise SystemExit("expected one positive integer n")

    n = int(data[0])
    maximum, grid = solve(n)
    print(f"maximum sum = {maximum}")
    print("arrangement:")
    print(*grid, sep="\n")


if __name__ == "__main__":
    main()

File: two_values_solution.cpp
#include <algorithm>
#include <cstdint>
#include <iostream>
#include <limits>
#include <stdexcept>
#include <string>
#include <vector>

using namespace std;

// Frontier state of a processed cell:
//   0: the cell contains 1
//   1: the cell contains 2 and is already adjacent to a 1
//   2: the cell contains 2 but still needs a 1 on its right or below

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int n;
    cin >> n;

    if (n <= 0) {
        cerr << "n must be positive\n";
        return 1;
    }

    vector<int> power3(n + 1, 1);
    for (int i = 1; i <= n; ++i) {
        if (power3[i - 1] > numeric_limits<int>::max() / 3) {
            cerr << "n is too large for this exact exponential algorithm\n";
            return 1;
        }
        power3[i] = power3[i - 1] * 3;
    }

    const int state_count = power3[n];
    const int INF = numeric_limits<int>::max() / 4;
    vector<int> dp(state_count, INF), next_dp(state_count, INF);

    // For every processed cell and resulting state, store
    //   (previous_state << 1) | (value - 1).
    // The low bit distinguishes value 1 from value 2.
    const uint32_t NO_PARENT = numeric_limits<uint32_t>::max();
    vector<vector<uint32_t>> parents;
    parents.reserve(n * n);

    // Before the first row there are no cells above it. State 1 is a neutral
    // virtual state: it neither requires the current cell to be 1 nor
    // falsely dominates it.
    int initial_state = 0;
    for (int c = 0; c < n; ++c) {
        initial_state += power3[c];
    }
    dp[initial_state] = 0;

    for (int r = 0; r < n; ++r) {
        for (int c = 0; c < n; ++c) {
            fill(next_dp.begin(), next_dp.end(), INF);
            vector<uint32_t> step_parent(state_count, NO_PARENT);
            const int current_power = power3[c];
            const int left_power = (c == 0 ? 0 : power3[c - 1]);

            for (int state = 0; state < state_count; ++state) {
                if (dp[state] == INF) {
                    continue;
                }

                const int above = (state / current_power) % 3;
                const int left = (c == 0 ? 1 : (state / left_power) % 3);

                // Put 1 in the current cell. It dominates the cell above,
                // the cell to the left, and future cells right/below it.
                int next_state = state - above * current_power;
                if (c > 0 && left == 2) {
                    // The previously undominated left cell is now dominated.
                    next_state -= left_power;
                }
                if (dp[state] + 1 < next_dp[next_state]) {
                    next_dp[next_state] = dp[state] + 1;
                    step_parent[next_state] =
                        (static_cast<uint32_t>(state) << 1) | 0U;
                }

                // Put 2 in the current cell. If the cell above still needs
                // domination, this choice is impossible because none of its
                // future neighbors remain.
                if (above != 2) {
                    const bool dominated = (above == 0) || (c > 0 && left == 0);
                    const int new_digit = dominated ? 1 : 2;
                    next_state = state + (new_digit - above) * current_power;
                    if (dp[state] < next_dp[next_state]) {
                        next_dp[next_state] = dp[state];
                        step_parent[next_state] =
                            (static_cast<uint32_t>(state) << 1) | 1U;
                    }
                }
            }

            dp.swap(next_dp);
            parents.push_back(move(step_parent));
        }
    }

    // After the last row, no frontier cell may still be waiting for a 1
    // below it.
    int minimum_ones = INF;
    int best_state = -1;
    for (int state = 0; state < state_count; ++state) {
        if (dp[state] == INF) {
            continue;
        }

        int encoded = state;
        bool valid = true;
        for (int c = 0; c < n; ++c) {
            if (encoded % 3 == 2) {
                valid = false;
                break;
            }
            encoded /= 3;
        }

        if (valid && dp[state] < minimum_ones) {
            minimum_ones = dp[state];
            best_state = state;
        }
    }

    // Follow the stored transitions backward to reconstruct one optimum.
    vector<string> grid(n, string(n, '?'));
    int state = best_state;
    for (int step = n * n - 1; step >= 0; --step) {
        uint32_t code = parents[step][state];
        int value = static_cast<int>(code & 1U) + 1;
        grid[step / n][step % n] = static_cast<char>('0' + value);
        state = static_cast<int>(code >> 1);
    }

    const long long maximum_sum = 2LL * n * n - minimum_ones;
    cout << "maximum sum = " << maximum_sum << '\n';
    cout << "arrangement:\n";
    for (const string& row : grid) {
        cout << row << '\n';
    }
    return 0;
}
